### PR TITLE
Add param widget ros2

### PR DESCRIPTION
**Pull request by kota-rr**
Ros2 version of parameter widget.

### CODE REVIEW

**Review by ardiya — APPROVED** [review]

File: cpp/notebooks/jupyter_ros_utils/ros2_jupyter.hpp [hunk 1]
```diff
@@ -160,39 +285,44 @@ class JupyterNodeAdvanced: ROS2Init, public NodeType {
...
         disp_box.add(spin_btn);
         disp_box.add(ros_valid);
         disp_box.display();
-        
+
         setup_thread();
     }
 
+
+public:
```
Comment (kunaltyagi) [inline cpp/notebooks/jupyter_ros_utils/ros2_jupyter.hpp:321]: Redundant public?

File: cpp/notebooks/jupyter_ros_utils/ros2_jupyter.hpp [hunk 2]
```diff
@@ -131,22 +134,144 @@ class ZeusSubscriber<MsgT, SubT, CallbackT, AllocatorT, MessageMemoryStrategyT,
...
+            if (nh->has_parameter(txt_param.value)) {
+                valid.value = true;
+                std::string tmp;
+                nh->get_parameter(txt_param.value, tmp);
+                txt_value.value = tmp;
+            } else {
+                valid.value = false;
+                txt_value.value = "";
+            }
+            usleep(200'000);
```
Comment (kunaltyagi) [inline cpp/notebooks/jupyter_ros_utils/ros2_jupyter.hpp:247]: Use rclcpp::Rate?

File: cpp/notebooks/jupyter_ros_utils/ros2_jupyter.hpp [hunk 3]
```diff
@@ -160,39 +285,44 @@ class JupyterNodeAdvanced: ROS2Init, public NodeType {
...
         disp_box.add(ros_valid);
         disp_box.display();
-        
+
         setup_thread();
     }
 
+
+public:
+    auto create_param_widgets() {
```
Comment (kunaltyagi) [inline cpp/notebooks/jupyter_ros_utils/ros2_jupyter.hpp:322]: Can we follow the ROS2 format for this? Ad 2 functions:
* set_parameter
* get_parameter
They create the same widget, but in different states (get_param widget is already spinning, while set_param widget is waiting for user input)

**Review by kunaltyagi — CHANGES_REQUESTED** [review]
@Apurv354 Since you'll be one of the people working on ROS2, please take over the PR from @kota-rr

File: cpp/notebooks/jupyter_ros_utils/ros2_jupyter.hpp (hunk 3 above)
Comment (kota-rr) [reply cpp/notebooks/jupyter_ros_utils/ros2_jupyter.hpp:322]: what is it supposed to do here? `JupyterNodeAdvanced` is already an Ros2 node, and already has `get_parameters` / `set_parameters`.

Comment (kunaltyagi) [reply cpp/notebooks/jupyter_ros_utils/ros2_jupyter.hpp:325]: Can we overload the functions since they are not virtual?

Comment (kota-rr) [reply cpp/notebooks/jupyter_ros_utils/ros2_jupyter.hpp:325]: Not possible directly to use the ros2 parameter code.
return types of `get_parameter` of `rcldpp::Node` is  `rclcpp::Parameter`
https://docs.ros2.org/bouncy/api/rclcpp/classrclcpp_1_1_node.html#a0bf1077fba623d72fe1b49805f6c0a5a
but `widget` also needs to be returned.

File: cpp/notebooks/jupyter_ros_utils/ros2_jupyter.hpp [hunk 4]
```diff
@@ -131,22 +137,146 @@ class ZeusSubscriber<MsgT, SubT, CallbackT, AllocatorT, MessageMemoryStrategyT,
...
+        }
+    }
+
+    void stop_thread() {
+        thr_running = false;
+        txt_param.disabled = false;
+        txt_value.disabled = false;
+        btn_wait.description = "Start Waiting";
+
+        usleep(1'000);
```
Comment (kunaltyagi) [inline cpp/notebooks/jupyter_ros_utils/ros2_jupyter.hpp:262]: this thread sleep instead?

http://en.cppreference.com/w/cpp/thread/sleep_for

File: cpp/notebooks/ros2_publisher.ipynb [hunk 5]
```diff
@@ -1,5 +1,13 @@
 {
  "cells": [
+  {
+   "cell_type": "markdown",
+   "id": "cc316365",
+   "metadata": {},
+   "source": [
+    "## Tutorial to use ros2 publisher"
```
Comment (kunaltyagi) [inline cpp/notebooks/ros2_publisher.ipynb:8]: Remove changes in this file please

File: cpp/notebooks/ros2_subscribers.ipynb [hunk 6]
```diff
@@ -1,9 +1,17 @@
 {
  "cells": [
+  {
+   "cell_type": "markdown",
+   "id": "0572b1ad",
+   "metadata": {},
+   "source": [
+    "## Tutorial to use ros2 subscriber"
```
Comment (kunaltyagi) [inline cpp/notebooks/ros2_subscribers.ipynb:8]: Same here: let apurv's pr take care of these. @Apurv354 Please check if these changes need to be added in your PR

**Review by kunaltyagi — CHANGES_REQUESTED** [review]
LGTM, except the comments

File: cpp/notebooks/ros2_subscribers.ipynb (hunk 6 above)
Comment (kota-rr) [reply cpp/notebooks/ros2_subscribers.ipynb:8]: the changes were added to make the formats of all ros2 examples to be consistent. But, yes, I will remove these for @Apurv354 to take care of these.

**Review by kunaltyagi — APPROVED** [review]

**Comment by kunaltyagi** [comment]
Merge?

**Comment by ardiya** [comment]
LGTM, let's merge the ros2param widget

**Pull request merged**